gitlab/utils.py

```diff
@@ -47,6 +49,39 @@ def response_content(
     return None
 
 
+def _transform_types(
+    data: Dict[str, Any], custom_types: dict, transform_files: Optional[bool] = True
+) -> Tuple[dict, dict]:
+    """Copy the data dict with attributes that have custom types and transform them
+    before being sent to the server.
+
+    If ``transform_files`` is ``True`` (default), also populates the ``files`` dict for
+    FileAttribute types with tuples to prepare fields for requests' MultipartEncoder:
+    https://toolbelt.readthedocs.io/en/latest/user.html#multipart-form-data-encoder
+
+    Returns:
+        A tuple of the transformed data dict and files dict"""
+
+    # Duplicate data to avoid messing with what the user sent us
+    data = data.copy()
+    files = {}
+
+    for attr_name, type_cls in custom_types.items():
+        if attr_name not in data.keys():
+            continue
+
+        type_obj = type_cls(data[attr_name])
+
+        # if the type if FileAttribute we need to pass the data as file
+        if transform_files and isinstance(type_obj, types.FileAttribute):
+            key = type_obj.get_file_name(attr_name)
+            files[attr_name] = (key, data.pop(attr_name))
```

Comment:
Using a `dataclass` (or even a `NamedTuple`) for the value, instead of a Tuple would really improve readability in my opinion. I have an example above called `FileInfoOrSomeBetterName`

What do you think?